uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/UserTokenGrantIT.java

```diff
@@ -0,0 +1,190 @@
+  final String user_token_secret = "secret";
+  final String user_token_public_id = "oauth_showcase_user_token_public";
+  final String empty_string = "";
+
+  @After
+  public void logout_and_clear_cookies() {
+    try {
+      webDriver.get(baseUrl + "/logout.do");
+    } catch (org.openqa.selenium.TimeoutException x) {
+      //try again - this should not be happening - 20 second timeouts
+      webDriver.get(baseUrl + "/logout.do");
+    }
+    webDriver.get(appUrl + "/j_spring_security_logout");
+    webDriver.manage().deleteAllCookies();
+  }
+
+  @Test
+  public void testExchangeFromConfidentialClientWithCfClientWithEmptySecret() {
+    // Given Create password token from confidential client
+    String token = getPasswordGrantToken(user_token_id, user_token_secret);
+
+    // When do user_token grant flow using public cf client (public, because of empty secret)
+    String newToken = doUserTokenGrant("cf", token, HttpStatus.OK);
+
+    // Then validation expected result
+    assertNotNull(newToken);
+    checkRefreshToken(newToken);
+  }
+
+  @Test
+  public void testExchangeFromConfidentialClientWithConfidentialClient() {
+    // Given Create password token from confidential client
+    String token = getPasswordGrantToken(user_token_id, user_token_secret);
+
+    // When do user_token grant flow using confidential oauth_showcase_user_token client
+    String newToken = doUserTokenGrant(user_token_id, token, HttpStatus.OK);
+
+    // Then validation expected result
+    assertNotNull(newToken);
+    checkRefreshToken(newToken);
+  }
+
+  @Test
+  public void testExchangeFromPublicClientWithPublicClient() {
+    // Given Create password token from public client
+    String token = getPasswordGrantToken(user_token_public_id, empty_string);
+
+    // When do user_token grant flow using public client
+    String newToken = doUserTokenGrant(user_token_public_id, token, HttpStatus.OK);
+
+    // Then validation expected result
+    assertNotNull(newToken);
+    checkRefreshToken(newToken);
+  }
+
+  @Test
+  public void testExchangeFromPublicClientWithConfidentialClient() {
+    // Given Create password token from public client
+    String token = getPasswordGrantToken(user_token_public_id, empty_string);
+
+    // When do user_token grant flow using confidential oauth_showcase_user_token client
+    String newToken = doUserTokenGrant(user_token_id, token, HttpStatus.OK);
+
+    // Then validation expected result
+    assertNotNull(newToken);
+    checkRefreshToken(newToken);
+  }
+
+  @Test
+  public void testExchangeFromConfidentialClientWithAdminClientExpectUnauthorized() {
+    // Given Create password token from public client
+    String token = getPasswordGrantToken(user_token_id, user_token_secret);
+
+    // When do user_token grant flow using admin client
+    String newToken = doUserTokenGrant("admin", token, HttpStatus.UNAUTHORIZED);
+
+    // Then validation expected result
+    assertNull(newToken);
```

Comment:
This assertNull() is only testing functionality in the test function doUserTokenGrant(). Not really needed here.